src/Search.f90

```diff
@@ -0,0 +1,286 @@
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+!                          Futility Development Group                          !
+!                             All rights reserved.                             !
+!                                                                              !
+! Futility is a jointly-maintained, open-source project between the University !
+! of Michigan and Oak Ridge National Laboratory.  The copyright and license    !
+! can be found in LICENSE.txt in the head directory of this repository.        !
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+!> @brief A Fortran 2003 module implementing some basic search algorithms.
+!>
+!> @par Module Dependencies
+!>  - @ref IntrType "IntrType": @copybrief IntrType
+!>
+!> @author Mitchell Young
+!>    @date 3/30/2017
+!>
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+MODULE Search
+  USE IntrType
+  USE Strings
+  IMPLICIT NONE
+
+  PRIVATE
+
+  PUBLIC :: upperBound
+  PUBLIC :: lowerBound
+  PUBLIC :: find
+  PUBLIC :: binarySearch
+
+  !> @brief Return the index of the first element of a list that compares
+  !> greater than the passed value
+  !> @param list the list to search. The list should be sorted, otherwise
+  !> undefined behavior.
+  !> @param val the value to find
+  !> @param tol the comparison tolerance to use for floating-point
+  !> implementations. Optional
+  !> @returns the index to the first element of the list that compares greater
+  !> than the passed search value
+  !>
+  !> If the value is greater than or equal to the largest element of the list,
+  !> returns SIZE + 1.
+  INTERFACE upperBound
+    MODULE PROCEDURE upper_bound_i
+    MODULE PROCEDURE upper_bound_r
+  ENDINTERFACE upperBound
+
+  !> @brief  Return the index of the first element of a list that does not
+  !> compare less than the passed value
+  !> @param list the list to search. The list should be sorted, otherwise
+  !> undefined behavior
+  !> @param val the value to find
+  !> @param tol the comparison tolerance to use for floating-point
+  !> implementations. Optional
+  !> @returns the index to the first element of the list that does not compare
+  !> less than the passed value
+  !>
+  !> If the value is greater than the last element of the list, returns SIZE+1
```

Comment:
This is correct. lower bound will return the index of the last element if the search value equal the last element. upper_bound wants to find the first element that is greater, therefore it will not find an element if the search value is equal to the last element in the sequence